core/gui/src/app/workspace/component/code-editor-dialog/code-editor.component.ts

```diff
@@ -169,11 +180,208 @@ export class CodeEditorComponent implements AfterViewInit, SafeStyle, OnDestroy
+            declineButton.addEventListener("click", declineListener, { once: true });
+          } else {
+            console.error("Buttons not found!");
+          }
+        }, 0);
+      });
+    });
+  }
+
+  // Called when the user clicks the "accept" button
+  public acceptCurrentAnnotation(): void {
+    // Avoid accidental calls
+    if (!this.showAnnotationSuggestion || !this.currentRange || !this.currentSuggestion) {
+      return;
+    }
+
+    if (this.currentRange && this.currentSuggestion) {
+      const selection = new monaco.Selection(
+        this.currentRange.startLineNumber,
+        this.currentRange.startColumn,
+        this.currentRange.endLineNumber,
+        this.currentRange.endColumn
+      );
+      this.insertTypeAnnotations(this.editor, selection, this.currentSuggestion);
+    }
+    // close the UI after adding the annotation
+    this.showAnnotationSuggestion = false;
+  }
+
+  // Called when the user clicks the "decline" button
+  public rejectCurrentAnnotation(): void {
+    // Do nothing except for closing the UI
+    this.showAnnotationSuggestion = false;
+  }
+
+  // Add the type annotation into monaco editor
+  private insertTypeAnnotations(
+    editor: monaco.editor.IStandaloneCodeEditor,
+    selection: monaco.Selection,
+    annotations: string
+  ) {
+    const endLineNumber = selection.endLineNumber;
+    const endColumn = selection.endColumn;
+    const range = new monaco.Range(
+      // Insert the content to the end of the selected code
+      endLineNumber,
+      endColumn,
+      endLineNumber,
+      endColumn
+    );
+    const text = `${annotations}`;
+    const op = {
+      range: range,
+      text: text,
+      forceMoveMarkers: true,
+    };
+    editor.executeEdits("add annotation", [op]);
+  }
+
+  // Add all necessary modules for type annotation at the first line of the Python UDF
+  private addAnnotationModule(editor: monaco.editor.IStandaloneCodeEditor) {
+    const model = editor.getModel();
+    if (!model) {
+      return;
+    }
+    const allCode = model.getValue();
+    const typingImports = [
+      "Any",
+      "Awaitable",
+      "Callable",
+      "Coroutine",
+      "Dict",
+      "FrozenSet",
+      "Generator",
+      "Generic",
+      "Iterable",
+      "Iterator",
+      "List",
+      "Mapping",
+      "Optional",
+      "Sequence",
+      "Set",
+      "Tuple",
+      "Type",
+      "TypeVar",
+      "Union",
+      "Deque",
+      "NamedTuple",
+      "TypedDict",
+      "Protocol",
+      "Literal",
+      "NewType",
+      "NoReturn",
+    ];
+    const importStatement = `from typing import (\n  ${typingImports.join(",\n  ")}\n)`;
+    if (!allCode.includes(importStatement)) {
+      const importOp = {
+        // Add the module at the first line
+        range: new monaco.Range(1, 1, 1, 1),
+        text: `${importStatement}\n\n`,
+      };
+      editor.executeEdits("add module", [importOp]);
+    }
+  }
```

Comment:
are we trying to modify user's import statements? I don't know the purpose of this function...

Reply:
Yes, in my test case the llm will generate some type annotation such as `List[string]`. If we don't have this module, there will be a error in the code. So we just want to import these typing modules for the generated Type annotation.

Reply:
then we can import them in the test cases explicitly. Let's avoid modifying the user's code (either appending or removing from it), as that could cause issues, especially when you are touching the imports.

Reply:
oh I take it back. so you are saying that you need those imports to support the generated annotations. Then I see the picture. To avoid changing user's code massively, can you temporarily use `from typing import *`?

Reply:
Do you mean we do not want to modify the user's code? If so, the user need to import the module by themselves if the LLM return the type annotation which need these modules.

Reply:
I will remove it from the code.

Reply:
Sorry I asked you to remove it from user's code. I am now thinking it is necessary to have those imports. but we can justs expose it from pytexera module (like what we did for `Iterator` and `Optional`). Please see if you can add other typing related imports into `https://github.com/Texera/texera/blob/master/core/amber/src/main/python/pytexera/__init__.py`

Reply:
Sure. I will do it tomorrow. All the other conversations are being resolved. I will let you know when I finish this last question.